File: geeksgforgeeks/challenge9_eggDropping/EggDropping.java
/*
    Created by Avinash on 6th November, 2017
*/
import java.io.*;
import java.util.*;
public class EggDropping
{
	public static int output;
	public static void main(String[] args)
	{
		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
		int n,k,threshold = 0;
		try
		{
		System.out.println("Enter number of eggs : ");
		n = Integer.parseInt(read.readLine());
		System.out.print("Enter number of floors : ");
		k = Integer.parseInt(read.readLine());
		Helper link = new Helper();	
		threshold = link.eggsThresholdFloorCache(n,k);
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
		System.out.print("The threshold floor for eggs is "+threshold);
	}
}

class Helper
{
	public int eggsThresholdFloor(int eggs, int floors)
	{
		if(floors == 0 || floors == 1)
        		return floors;
        	if(eggs == 1)
        		return floors;
       		int min = Integer.MAX_VALUE, threshold;
      		for(int k = 1; k <= floors; k++)
        	{
        		threshold = Math.max(eggsThresholdFloor(eggs-1, k-1), eggsThresholdFloor(eggs, floors - k));
        		if(threshold < min)
        			min = threshold;
        	}
        	return min + 1;
        		
	}

	public int eggsThresholdFloorCache(int eggs, int floors)
	{
		int cache[][] = new int[eggs+1][floors+1];
		int temp;
		for(int i = 1; i <= eggs; i++)
		{
			cache[i][0] = 0;
			cache[i][1] = 1;	
	        }
	        for(int i = 1; i <= floors; i++)
	        {
	        	cache[1][i] = i;
	        }
	        for(int i = 2; i <= eggs; i++)
	        {
	        	for(int j = 2; j <= floors; j++)
	        	{
	        		cache[i][j] = Integer.MAX_VALUE;
	        		for(int x = 1; x <= j; x++)
	        		{
	        			temp = 1 + Math.max(cache[i-1][x-1], cache[i][j-x]);
	        			if(temp < cache[i][j])
	        				cache[i][j] = temp;
	        		}
	        	}
	        }
	        return cache[eggs][floors];	
	        	
	}
  	
}	                    	
			

File: geeksgforgeeks/challenge9_eggDropping/eggDropping.py
# Created by Avinash on 6th Nov, 2017
import sys
def process():
	n = int(input("Enter the number of eggs : "))
	k = int(input("Enter the number of floors : "))
	threshold = eggsThresholdFloorCache(n,k)
	print("The threshold floor for given eggs is %-5d" %threshold)

def eggsThresholdFloor(n,k):
	if k == 0 or k == 1 :
		return k
	if n == 1 :
		return k
	min = sys.maxsize
	for j in range(1,k + 1) :
		threshold = max(eggsThresholdFloor(n - 1, j - 1),eggsThresholdFloor(n,k - j))
		if(threshold < min) :
			min = threshold
	return min + 1

def  eggsThresholdFloorCache(n,k) :
	threshold = [[0 for x in range(k+1)]for x in range(n+1)]
	for i in range(1, n+1):
		threshold[i][1] = 1
		threshold[i][0] = 0
	for j in range(1, k+1) :
		threshold[1][j] = j
	for i in range(2, n+1):
		for j in range(2, k+1):
			threshold[i][j] = sys.maxsize
			for x in range(1,j+1) :
				res = 1 + max(threshold[i-1][x-1],threshold[i][j-x])
				if res < threshold[i][j]:
					threshold[i][j] = res
	return threshold[n][k]	
					

if __name__=="__main__":
	process()
		
			